Flaky kv/memberlist TestZoneAwareRouting_EndToEnd

Can be reproduced fairly reliably with:

```
go test -failfast -count=100 -run=TestZoneAwareRouting_EndToEnd ./kv/memberlist/...
```

```
--- FAIL: TestZoneAwareRouting_EndToEnd (2.44s)
    --- FAIL: TestZoneAwareRouting_EndToEnd/state_sync_via_push/pull (2.23s)
        node_zone_aware_routing_test.go:335: waiting for node zone-a-bridge-1b7b6564 in zone zone-a with role bridge to see all other nodes
        node_zone_aware_routing_test.go:335: waiting for node zone-a-member-31e0cf91 in zone zone-a with role member to see all other nodes
        node_zone_aware_routing_test.go:335: waiting for node zone-a-member-6d7763f0 in zone zone-a with role member to see all other nodes
        node_zone_aware_routing_test.go:335: waiting for node zone-b-bridge-43264a7f in zone zone-b with role bridge to see all other nodes
        node_zone_aware_routing_test.go:335: waiting for node zone-b-member-e517e41f in zone zone-b with role member to see all other nodes
        node_zone_aware_routing_test.go:336: expected 6, got 4
FAIL
FAIL    github.com/grafana/dskit/kv/memberlist  51.633s
FAIL
```

A subset of the output from the test:

```
instance=member-zone-b-1 level=debug msg="Initiating push/pull sync with: zone-b-member-61c84978 127.0.0.1:42415"
instance=member-zone-b-2 level=debug msg="Stream connection from=127.0.0.1:57758"
instance=member-zone-a-1 level=debug msg="Stream connection from=127.0.0.1:55100"
instance=member-zone-a-2 level=debug msg="Initiating push/pull sync with: zone-a-member-31e0cf91 127.0.0.1:33037"
level=info instance=member-zone-b-2 msg="leaving memberlist cluster"
level=info instance=member-zone-a-1 msg="leaving memberlist cluster"
level=info instance=member-zone-a-2 msg="leaving memberlist cluster"
level=info instance=member-zone-b-1 msg="leaving memberlist cluster"
level=info instance=bridge-zone-b msg="leaving memberlist cluster"
instance=member-zone-a-1 level=debug msg="Initiating push/pull sync with: zone-a-bridge-1b7b6564 127.0.0.1:38951"
instance=member-zone-b-2 level=debug msg="Initiating push/pull sync with: zone-b-member-e517e41f 127.0.0.1:43609"
instance=bridge-zone-a level=debug msg="Stream connection from=127.0.0.1:36744"
instance=member-zone-b-1 level=debug msg="Stream connection from=127.0.0.1:39806"
instance=bridge-zone-a level=debug msg="Initiating push/pull sync with: zone-b-bridge-43264a7f 127.0.0.1:36399"
instance=bridge-zone-b level=debug msg="Stream connection from=127.0.0.1:49612"
instance=bridge-zone-b level=debug msg="Initiating push/pull sync with: zone-a-bridge-1b7b6564 127.0.0.1:38951"
instance=bridge-zone-a level=debug msg="Stream connection from=127.0.0.1:36750"
instance=member-zone-a-2 level=debug msg="Initiating push/pull sync with: zone-a-bridge-1b7b6564 127.0.0.1:38951"
instance=bridge-zone-a level=debug msg="Stream connection from=127.0.0.1:36760"
instance=member-zone-a-1 level=debug msg="Initiating push/pull sync with: zone-a-bridge-1b7b6564 127.0.0.1:38951"
instance=bridge-zone-a level=debug msg="Stream connection from=127.0.0.1:36770"
instance=bridge-zone-b level=debug msg="Initiating push/pull sync with: zone-a-bridge-1b7b6564 127.0.0.1:38951"
instance=bridge-zone-a level=debug msg="Stream connection from=127.0.0.1:36772"
instance=member-zone-a-2 level=debug msg="Initiating push/pull sync with: zone-a-bridge-1b7b6564 127.0.0.1:38951"
instance=bridge-zone-a level=debug msg="Stream connection from=127.0.0.1:36774"
level=error instance=member-zone-b-1 msg="error when leaving memberlist cluster" err="timeout waiting for leave broadcast"
level=error instance=member-zone-a-1 msg="error when leaving memberlist cluster" err="timeout waiting for leave broadcast"
level=error instance=bridge-zone-b msg="error when leaving memberlist cluster" err="timeout waiting for leave broadcast"
level=error instance=member-zone-a-2 msg="error when leaving memberlist cluster" err="timeout waiting for leave broadcast"
level=error instance=member-zone-b-2 msg="error when leaving memberlist cluster" err="timeout waiting for leave broadcast"
level=info instance=bridge-zone-a msg="leaving memberlist cluster"
level=error instance=bridge-zone-a msg="error when leaving memberlist cluster" err="timeout waiting for leave broadcast"
```

--- a/kv/memberlist/node_zone_aware_routing_test.go
+++ b/kv/memberlist/node_zone_aware_routing_test.go
@@ -312,8 +312,8 @@ func TestZoneAwareRouting_EndToEnd(t *testing.T) {
 			memberB2 := NewKV(makeConfig(bridgeAAddr, "zone-b", "member"), log.WithPrefix(logger, "instance", "member-zone-b-2"), &staticDNSProviderMock{}, prometheus.NewPedanticRegistry())
 
 			// Start services in what we consider the worst case scenario:
-			// zone-b members first, so they don't see any zone-b bridge at startup,
-			// but they're also not allowed to initiate a push/pull to any zone-a member or bridge.
+			// zone-b members first, so they don't see any zone-b bridge at startup, but they're also
+			// not allowed to initiate a push/pull or gossip to any zone-a member or bridge.
 			zoneBMembers, err := services.NewManager(memberB1, memberB2)
 			require.NoError(t, err)
 			require.NoError(t, services.StartManagerAndAwaitHealthy(ctx, zoneBMembers))
@@ -321,7 +321,7 @@ func TestZoneAwareRouting_EndToEnd(t *testing.T) {
 				require.NoError(t, services.StopManagerAndAwaitStopped(ctx, zoneBMembers))
 			})
 
-			// The start other nodes.
+			// Then start other nodes.
 			otherMembers, err := services.NewManager(memberA1, memberA2, bridgeB)
 			require.NoError(t, err)
 			require.NoError(t, services.StartManagerAndAwaitHealthy(ctx, otherMembers))

